### PR TITLE
(#254) codegen: if else block

### LINKED ISSUE #254: codegen: if else

### PR DESCRIPTION
**Pull request by dannypsnl**
resolve #254

### CODE REVIEW

**Comment by dannypsnl** [comment]
@YanHerChen To review this PR, [elz wiki](https://github.com/dannypsnl/elz/wiki/LLVM-Notes) might help. The point is to ensure this PR work correctly and readable enough for you. The unit test could help you get the target of the code generator.

NOTE:
- `Rc` is a single-thread reference-counting pointer in Rust
- ir means: [Intermediate representation](https://en.wikipedia.org/wiki/Intermediate_representation)
- In ir, there only has no nested block
- **continues terminators** this situation is invalid in LLVM

File: src/codegen/ir.rs [hunk 1]
```diff
@@ -58,54 +59,127 @@ pub(crate) struct TypeDefinition {
     pub(crate) fields: Vec<Type>,
 }
 
+/// Label represents a location which can be the target of jump instructions
+///
+/// LLVM example: `; <label>:9:`, here `id` is `9`
```
Comment (dannypsnl) [inline src/codegen/ir.rs]: NOTE: Remove outdated comment

File: src/codegen/ir.rs [hunk 2]
```diff
@@ -58,54 +59,127 @@ pub(crate) struct TypeDefinition {
...
+        }
+    }
 }
 
 #[derive(Debug, Clone, PartialEq)]
 pub(crate) struct Body {
     // helper part
     pub(crate) counter: u64,
     // value part
-    pub(crate) statements: Vec<Instruction>,
```
Comment (dannypsnl) [inline src/codegen/ir.rs:72]: Reverse instructions and counter to remove this comment

File: src/codegen/ir.rs [hunk 3]
```diff
@@ -58,54 +59,127 @@ pub(crate) struct TypeDefinition {
...
     pub(crate) fn from_ast(b: &ast::Body, module: &Module) -> Body {
         let mut body = Body {
             counter: 0,
-            statements: vec![],
+            instructions: vec![],
         };
         match b {
             ast::Body::Expr(e) => {
-                body.statements = vec![Instruction::Return(Some(Expr::from_ast(e, module)))];
+                body.instructions = vec![Instruction::Return(Some(Expr::from_ast(e, module)))];
```
Comment (dannypsnl) [inline src/codegen/ir.rs]: instructions.push?

File: src/codegen/ir.rs [hunk 4]
```diff
@@ -58,54 +59,127 @@ pub(crate) struct TypeDefinition {
...
-                        instruction
+                    };
+                    self.instructions.push(inst)
+                }
+                Expression(expr) => {
+                    let inst = Instruction::TempVariable(
+                        self.get_and_update_count(),
+                        Expr::from_ast(expr, module),
+                    );
+                    self.instructions.push(inst)
```
Comment (dannypsnl) [inline src/codegen/ir.rs]: Remove temp variable: inst?

File: src/codegen/ir.rs [hunk 5]
```diff
@@ -58,54 +59,127 @@ pub(crate) struct TypeDefinition {
...
+            }
+        }
+    }
+    fn get_and_update_count(&mut self) -> u64 {
+        let tmp = self.counter;
+        self.counter += 1;
+        tmp
+    }
+    fn goto(&mut self, label: &Rc<Label>) {
+        if let Some(inst) = self.instructions.last() {
```
Comment (dannypsnl) [inline src/codegen/ir.rs]: Maybe we should judge duplicate terminator case by case?

File: src/codegen/llvm.rs [hunk 6]
```diff
@@ -57,15 +57,40 @@ impl LLVMValue for ir::Instruction {
...
+            Goto(block) => format!("br {}", block.llvm_represent()),
+            Label(label) => format!("label{}:", label.id),
         }
     }
 }
 
+impl LLVMValue for ir::Label {
+    fn llvm_represent(&self) -> String {
+        format!("label %label{}", self.id)
+    }
```
Comment (dannypsnl) [inline src/codegen/llvm.rs:85]: Probably introduce a label.name() to keep ”label<id>” consistently?

File: src/codegen/llvm.rs [hunk 7]
```diff
@@ -57,15 +57,40 @@ impl LLVMValue for ir::Instruction {
...
+}
+
 impl LLVMValue for ir::Body {
     fn llvm_represent(&self) -> String {
         let mut s = String::new();
-        for stmt in &self.statements {
-            s.push_str(format!("  {}\n", stmt.llvm_represent()).as_str());
+        for instruction in &self.instructions {
+            match instruction {
+                ir::Instruction::Label(_) => {
```
Comment (YanHerChen) [inline src/codegen/llvm.rs]: probably replace _ with ..

**Review by YanHerChen — CHANGES_REQUESTED** [review]

**Comment by dannypsnl** [comment]
@YanHerChen PR updated! Please review again.

**Review by YanHerChen — APPROVED** [review]

**Pull request merged**